git(-with-openssh): Improve registry srcipt & relevant logic, update checkver & autoupdate

SorYoshino opened the pull request:
### Summary

Improves the `git` and `git-with-openssh` manifests by refining metadata, restructuring post-install behavior, modernizing registry script handling, and aligning both packages with Scoop manifest conventions.

### Related issues or pull requests

- Relates to #7061 
- Relates to #3685 
- Relates to #6259 

### Changes

- Refine `description` to a formal and complete wording  
- Update `license` to `GPL-2.0-or-later` and added license URL  
- Refactor `notes` section using formal language and clear execution instructions  
- Reorganize `bin` entries
- Add `env_add_path` for automatic PATH integration  
- Standardize `shortcuts` names (grouped under “Git\*”)   
- Simplify and generalize `post_install` scripts
- Add uninstall script blocks to restore registry changes on removal  
- update checkver & autoupdate  
- Normalize registry file generation encoding and escaping rules  

### Notes

- The official installer directly appends the `cmd` directory to the `Path` environment variable, and the useful tools contained in this directory extend well beyond `git`, `gitk`, and `scalar`. Rather than enumerating these utilities individually in the `bin` field, it is more appropriate to add the entire `cmd` directory to `Path`.
- Tools related to `gpg` are not in the same category as `ssh`; however, they are critical during development and should be included in the `bin` field for both `git` and `git-with-openssh`.
- The contents of `install-context.reg` and `install-associations.reg` have been realigned with the behavior of the official installer, and the associated logic has been refined accordingly.

### Testing

<details>
<summary>The test results are as follows:</summary><br>

```powershell
┏[ D:\Software\Scoop\Local\apps\scoop\current\bin][ master ≡]
└─> .\checkver.ps1 -App git -Dir 'D:\Temporary\Software\Microsoft\Windows Sandbox\Repositories\Scoop\Buckets\Main\bucket' -f
git: 2.52.0 (scoop version is 2.52.0)
Forcing autoupdate!
Autoupdating git
DEBUG[1766751456] $substitutions (hashtable) -> D:\Software\Scoop\Local\apps\scoop\current\lib\autoupdate.ps1:221:5
DEBUG[1766751456] $substitutions.$dotVersion                    2.52.0
DEBUG[1766751456] $substitutions.$match1                        2.52.0
DEBUG[1766751456] $substitutions.$basenameNoExt                 PortableGit-2.52.0-arm64.7z
DEBUG[1766751456] $substitutions.$urlNoExt                      https://github.com/git-for-windows/git/releases/download/v2.52.0.windows.1/PortableGit-2.52.0-arm64.7z
DEBUG[1766751456] $substitutions.$url                           https://github.com/git-for-windows/git/releases/download/v2.52.0.windows.1/PortableGit-2.52.0-arm64.7z.exe
DEBUG[1766751456] $substitutions.$patchVersion                  0
DEBUG[1766751456] $substitutions.$majorVersion                  2
DEBUG[1766751456] $substitutions.$version                       2.52.0
DEBUG[1766751456] $substitutions.$matchTag                      v2.52.0.windows.1
DEBUG[1766751456] $substitutions.$minorVersion                  52
DEBUG[1766751456] $substitutions.$baseurl                       https://github.com/git-for-windows/git/releases/download/v2.52.0.windows.1
DEBUG[1766751456] $substitutions.$matchHead                     2.52.0
DEBUG[1766751456] $substitutions.$matchTail
DEBUG[1766751456] $substitutions.$preReleaseVersion             2.52.0
DEBUG[1766751456] $substitutions.$dashVersion                   2-52-0
DEBUG[1766751456] $substitutions.$cleanVersion                  2520
DEBUG[1766751456] $substitutions.$buildVersion
DEBUG[1766751456] $substitutions.$underscoreVersion             2_52_0
DEBUG[1766751456] $substitutions.$basename                      PortableGit-2.52.0-arm64.7z.exe
DEBUG[1766751456] $hashfile_url = https://github.com/git-for-windows/git/releases/tag/v2.52.0.windows.1 -> D:\Software\Scoop\Local\apps\scoop\current\lib\autoupdate.ps1:224:5
Searching hash for PortableGit-2.52.0-arm64.7z.exe in https://github.com/git-for-windows/git/releases/tag/v2.52.0.windows.1
DEBUG[1766751456] $regex = (?s)PortableGit-2\.52\.0-arm64\.7z\.exe.*?([a-fA-F0-9]{64}) -> D:\Software\Scoop\Local\apps\scoop\current\lib\autoupdate.ps1:78:9
Found: bdc2884b321152225498fadc97cad68c244e047310ea686e4fe18ad7257e5b72 using Extract Mode
Writing updated git manifest

┏[ D:\Software\Scoop\Local\apps\scoop\current\bin][ master ≡]
└─> .\checkver.ps1 -App git-with-openssh -Dir 'D:\Temporary\Software\Microsoft\Windows Sandbox\Repositories\Scoop\Buckets\Main\bucket' -f
git-with-openssh: 2.52.0 (scoop version is 2.52.0)
Forcing autoupdate!
Autoupdating git-with-openssh
DEBUG[1766751481] $substitutions (hashtable) -> D:\Software\Scoop\Local\apps\scoop\current\lib\autoupdate.ps1:221:5
DEBUG[1766751481] $substitutions.$dotVersion                    2.52.0
DEBUG[1766751481] $substitutions.$match1                        2.52.0
DEBUG[1766751481] $substitutions.$basenameNoExt                 PortableGit-2.52.0-arm64.7z
DEBUG[1766751481] $substitutions.$urlNoExt                      https://github.com/git-for-windows/git/releases/download/v2.52.0.windows.1/PortableGit-2.52.0-arm64.7z
DEBUG[1766751481] $substitutions.$url                           https://github.com/git-for-windows/git/releases/download/v2.52.0.windows.1/PortableGit-2.52.0-arm64.7z.exe
DEBUG[1766751481] $substitutions.$patchVersion                  0
DEBUG[1766751481] $substitutions.$majorVersion                  2
DEBUG[1766751481] $substitutions.$version                       2.52.0
DEBUG[1766751481] $substitutions.$matchTag                      v2.52.0.windows.1
DEBUG[1766751481] $substitutions.$minorVersion                  52
DEBUG[1766751481] $substitutions.$baseurl                       https://github.com/git-for-windows/git/releases/download/v2.52.0.windows.1
DEBUG[1766751481] $substitutions.$matchHead                     2.52.0
DEBUG[1766751481] $substitutions.$matchTail
DEBUG[1766751481] $substitutions.$preReleaseVersion             2.52.0
DEBUG[1766751481] $substitutions.$dashVersion                   2-52-0
DEBUG[1766751481] $substitutions.$cleanVersion                  2520
DEBUG[1766751481] $substitutions.$buildVersion
DEBUG[1766751481] $substitutions.$underscoreVersion             2_52_0
DEBUG[1766751481] $substitutions.$basename                      PortableGit-2.52.0-arm64.7z.exe
DEBUG[1766751481] $hashfile_url = https://github.com/git-for-windows/git/releases/tag/v2.52.0.windows.1 -> D:\Software\Scoop\Local\apps\scoop\current\lib\autoupdate.ps1:224:5
Searching hash for PortableGit-2.52.0-arm64.7z.exe in https://github.com/git-for-windows/git/releases/tag/v2.52.0.windows.1
DEBUG[1766751481] $regex = (?s)PortableGit-2\.52\.0-arm64\.7z\.exe.*?([a-fA-F0-9]{64}) -> D:\Software\Scoop\Local\apps\scoop\current\lib\autoupdate.ps1:78:9
Found: bdc2884b321152225498fadc97cad68c244e047310ea686e4fe18ad7257e5b72 using Extract Mode
Writing updated git-with-openssh manifest

┏[ D:\Software\Scoop\Local\apps\scoop\current\bin][ master ≡]
└─> scoop install Unofficial/git
Installing 'git' (2.52.0) [64bit] from 'Unofficial' bucket
Loading PortableGit-2.52.0-64-bit.7z.exe from cache.
Checking hash of PortableGit-2.52.0-64-bit.7z.exe ... ok.
Extracting PortableGit-2.52.0-64-bit.7z.exe ... done.
Linking D:\Software\Scoop\Local\apps\git\current => D:\Software\Scoop\Local\apps\git\2.52.0
Creating shim for 'sh'.
Creating shim for 'git-bash'.
Making D:\Software\Scoop\Local\shims\git-bash.exe a GUI binary.
Creating shim for 'gpg'.
Creating shim for 'gpg-agent'.
Creating shim for 'gpgconf'.
Creating shim for 'gpg-connect-agent'.
Creating shim for 'pinentry'.
Making D:\Software\Scoop\Local\shims\pinentry.exe a GUI binary.
Creating shortcut for Git\Git Bash (git-bash.exe)
Creating shortcut for Git\Git CMD (git-cmd.exe)
Creating shortcut for Git\Git GUI (git-gui.exe)
Adding D:\Software\Scoop\Local\apps\git\current\cmd to your path.
Running post_install script..done.
'git' (2.52.0) was installed successfully!
Notes
-----
To register file associations, please execute the following command:
reg import "D:\Software\Scoop\Local\apps\git\current\install-associations.reg"

To register the context menu entry, please execute the following command:
reg import "D:\Software\Scoop\Local\apps\git\current\install-context.reg"

To set Git Credential Manager Core for portable Git, please execute the following command:
git config --global credential.helper manager

┏[ D:\Software\Scoop\Local\apps\scoop\current\bin][ master ≡]
└─> scoop install 'D:\Temporary\Software\Microsoft\Windows Sandbox\Repositories\Scoop\Buckets\Main\bucket\git-with-openssh.json'
Installing 'git-with-openssh' (2.52.0) [64bit] from 'Unofficial' bucket
Loading PortableGit-2.52.0-64-bit.7z.exe from cache.
Checking hash of PortableGit-2.52.0-64-bit.7z.exe ... ok.
Extracting PortableGit-2.52.0-64-bit.7z.exe ... done.
Linking D:\Software\Scoop\Local\apps\git-with-openssh\current => D:\Software\Scoop\Local\apps\git-with-openssh\2.52.0
Creating shim for 'sh'.
Creating shim for 'git-bash'.
Making D:\Software\Scoop\Local\shims\git-bash.exe a GUI binary.
Creating shim for 'gpg'.
Creating shim for 'gpg-agent'.
Creating shim for 'gpgconf'.
Creating shim for 'gpg-connect-agent'.
Creating shim for 'pinentry'.
Making D:\Software\Scoop\Local\shims\pinentry.exe a GUI binary.
Creating shim for 'scp'.
Creating shim for 'sftp'.
Creating shim for 'ssh'.
Creating shim for 'ssh-add'.
Creating shim for 'ssh-agent'.
Creating shim for 'ssh-copy-id'.
Creating shim for 'ssh-keygen'.
Creating shim for 'ssh-keyscan'.
Creating shim for 'ssh-pageant'.
Creating shortcut for Git\Git Bash (git-bash.exe)
Creating shortcut for Git\Git CMD (git-cmd.exe)
Creating shortcut for Git\Git GUI (git-gui.exe)
Adding D:\Software\Scoop\Local\apps\git-with-openssh\current\cmd to your path.
Running post_install script...done.
'git-with-openssh' (2.52.0) was installed successfully!
Notes
-----
To register file associations, please execute the following command:
reg import "D:\Software\Scoop\Local\apps\git-with-openssh\current\install-file-associations.reg"

To register the context menu entry, please execute the following command:
reg import "D:\Software\Scoop\Local\apps\git-with-openssh\current\install-context.reg"

To set Git Credential Manager Core for portable Git, please execute the following command:
git config --global credential.helper manager
```

</details>

- [x] Use conventional PR title: `<manifest-name[@version]|chore>: <general summary of the pull request>`
- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md)

Lutra-Fs commented on bucket/git-with-openssh.json:
```diff
@@ -1,14 +1,20 @@
     "homepage": "https://gitforwindows.org",
-    "license": "GPL-2.0-only",
+    "license": {
+        "identifier": "GPL-2.0-or-later",
```
wrong license. see https://github.com/git-for-windows/git/blob/00d4f5e7d9c7c4ea5d22e2f395bcce6e749f4672/COPYING#L3 which you attached as url:

>  Note that the only valid version of the GPL as far as this project
>  is concerned is _this_ particular version of the license (ie v2, not
>  v2.2 or v3.x or whatever), unless explicitly otherwise stated.

By the way, I’d prefer URLs are only used when a standard SPDX identifier doesn’t resolve the issue like “proprietary” or “NOASSERTION”.  It’s as if the SPDX can describe the licence.

SorYoshino replied:
> wrong license. see https://github.com/git-for-windows/git/blob/00d4f5e7d9c7c4ea5d22e2f395bcce6e749f4672/COPYING#L3 which you attached as url:
> 
> > Note that the only valid version of the GPL as far as this project
> > is concerned is _this_ particular version of the license (ie v2, not
> > v2.2 or v3.x or whatever), unless explicitly otherwise stated.

Right, thank you for reminding me.

> By the way, I’d prefer URLs are only used when a standard SPDX identifier doesn’t resolve the issue like “proprietary” or “NOASSERTION”. It’s as if the SPDX can describe the licence.

I think it would be better to directly specify the corresponding license URL, similar to Winget, so that each license can be explicitly matched.

SorYoshino commented:
Hello, @Lutra-Fs .

The previously mentioned license issue has been resolved. Are there any other concerns with this PR? Is there anything I can do to help expedite the review process?

Lutra-Fs commented on bucket/git.json:
```diff
@@ -21,60 +27,66 @@
-        "        }",
-        "        Set-Content -Path $reg_file_out -Value $content -Encoding Ascii",
-        "    }",
+        "$git_path = $dir -replace '\\\\', '\\\\'",
```
same with git-with-openssh

Lutra-Fs commented on bucket/git-with-openssh.json:
```diff
@@ -1,14 +1,20 @@
     "notes": [
-        "Set Git Credential Manager Core by running: \"git config --global credential.helper manager\"",
+        "To register file associations, please execute the following command:",
+        "reg import \"$dir\\install-file-associations.reg\"",
```
wrong file name referenced.

SorYoshino commented on bucket/git-with-openssh.json:
```diff
@@ -21,71 +27,79 @@
-        "        }",
-        "        Set-Content -Path $reg_file_out -Value $content -Encoding Ascii",
-        "    }",
+        "$git_path = $dir -replace '\\\\', '\\\\'",
```
Yes, the first argument for `-replace` is a regular expression, and the second argument is a string.

```powershell
┏[ ~]
└─> $env:PUBLIC
C:\Users\Public

┏[ ~]
└─> $env:PUBLIC -replace '\\', '\\'
C:\\Users\\Public
```

Regular expressions offer greater flexibility than direct string manipulation, hence my use of `-replace`.

Lutra-Fs replied:
Yes, yesterday I misread the expression due to the double escape. By the way, I still believe the current approach while flexible isn’t easy to read and understand. Can we stay with `$dir.Replace('\','\\')` for clarity？ Our goal is to manipulate the path, not make any other changes. We know there won’t be any non path separator backslashes in the path.

SorYoshino commented:
/verify

SorYoshino commented:
> If you are available, when addressing these comments, pls also submit your verification script, and results.

Thank you for reviewing this pull request.

All comments have been addressed. I have also rechecked all files containing changes and should be clear of any further issues.

SorYoshino commented:
Hello, @Lutra-Fs .

Sorry for the late reply.

- Since this PR has already been merged, should #6259 be closed? It only added `reg import` in the `notes` field.

- Also, when you have time, could you please review #7460? In addition to adding the `arm64` architecture, the other changes are similar to those in this PR, so it shouldn’t take too much of your time.
